### ISSUE TITLE
A clarification over the procedure of creating a clean train and test set from a noisy train/test split

### ISSUE THREAD

**Issue by filippoBUO** [cleanlab/cleanlab#227]
Hi,

I'm writing this issue because I'd like to know if I'm doing things right.

I want to create a clean train/test split and analyze the impact of the correction of noisy train labels in the test set accuracy.

I did this in order to create a clean test set:

1) I started with a train test split of my full noisy dataset. As a result I have a noisy train and a noisy test set.

2) I run cleanlab on the noisy train set to obtain a clean train set (7.5% of issues found). After this step I have a clean train set and a noisy test set.

3) I fit a classifier on the clean train set and then calculate out of sample probs on the noisy test set. Then I run cleanlab on the noisy test set to get a clean test set. As a result after this step I should have a clean train and clean test, right?

Note: At every step I'm correcting the labels issues suggested by cleanlab (I'm not eliminating them).

Ok, after all these steps I want to analyze the difference in accuracy on the clean test set in these two following cases:
      a) If I train a model on the original noisy train set.
      b) If I train a model on the clean train set.

My question is: it a fair procedure to look at the impact of correcting or not the noisy labels? Actually I found that for that particular level of noise (7.5%) there is no big difference in term of accuracy on the clean test set between correcting or no correcting those labels.

Thank you so much and congratulations for your awesome work!

**Comment by jwmueller** [comment]
Hi @filippoBUO 

Could you share a code snippet when you say: "I run cleanlab on the noisy train set to obtain a clean train set"
(just because there are multiple ways to go about this, in particular where did your `pred_probs` come from here?)

In fact it would be most helpful if you can share all the code for your steps 1-3 + (a), (b) as well as the outputs + accuracy values you're seeing. Just to be sure we're on the same page!

Overall this does seem like a fair procedure to look at the impact of correcting the noisy labels vs not. Some types of models may be more robust to label-noise than others, so there's not always guaranteed to be a big difference. But for most models I'd expect a nontrivial difference in this case with 7.5% labels being wrong, so I'm curious to learn more about your scenario!

**Comment by filippoBUO** [comment]
Hi @jwmueller 

Thank you for your answer.
I'm doing NLP and working with a binary dataset of about 390k rows. In step one I do a train/test split of my noisy dataset with the following proportions (99% / 1%, keeping the same proportion of the classes in both sets).

The model I'm using to calculate the out-of-sample probabilities in the train via cross-validation (cv I do it manually without using the cleanlab function) is a Logisitc Regression with an embedding layer is the following:

```
inputs = tf.keras.layers.Input(shape=(100,))
h = tf.keras.layers.Embedding(input_dim=10000, output_dim=16)(inputs)
h = tf.keras.layers.Lambda(lambda x: tf.keras.backend.mean(x, axis=1))(h)
outputs = tf.keras.layers.Dense(2, activation='softmax')(h)

model = tf.keras.Model(inputs=inputs, outputs=outputs, name="my_model")

model.compile(optimizer=tf.keras.optimizers.Adam(),
             loss=tf.keras.losses.sparse_categorical_crossentropy,
             metrics=["accuracy"])

model.fit(padded_train_seq, y_train, epochs=10, batch_size=64)
```

Cleanlab found 28k of labels issues (7.5% of the total) and the noise_matrix estimated is 
[[0.84432946, 0.04365889],
[0.15567054, 0.95634111]]
The issues found by cleanlab I check and in fact they are clear errors (as a matter of fact the last issue (ordered by normalized margin) returned by cleanlab has a self-confidence of 0.30). So I decided to flip the labels of these issues to the correct suggested class.
This is the end of step 2 of above.

Regarding step 3, I fit the same model as above on the clean train set obtained in step 2 (now without cv) and do the predict probs on the noisy test set. Again the issues found by clean lab (7.9% of issues found) on the test set are clear errors so I basically flip the labels of these issues to the correct suggested class as before. After step 3 I have a "clean" test dataset.

Now I have a clean train/test split and the original noisy train dataset. 
What I did is fit the same model as above on both the noisy train set (step a) and clean train set (step b), and test that model on the clean test set.
These are the results:
a) accuracy over noisy train: 0.8872, accuracy over clean test: 0.9669
b) accuracy over clean train: 0.9583, accuracy over clean test: 0.9649

Note: benchmark) accuracy over noisy train: 0.8872, accuracy over noisy test: 0.8880.

The only improvement in accuracy I see  between step a) and b) is the one on the train which is about 7% almost the same as the amount of labels issues found by cleanlab in the train (7.5%).

What puzzle me is that given a clean test set, it seems that the model is fairly robust despite of a 7.5% of noise in the data. 

Thank you so much!

**Comment by jwmueller** [comment]
Thanks for providing all these details, I'm glad to see cleanlab found real issues in your data!

I think your overall procedure was done properly, so this is mostly a fair comparison. Perhaps your model is simply very robust to the noise...  Here's a couple other things to consider:
 
 1) Maybe your test data is too small and these results are simply high-variance?  
Eg. it's a bit weird to see test accuracy higher than train accuracy.

Another metric you could quickly look at is:  train accuracy on clean labels achieved by model trained on noisy labels. If this is also similar to 0.9583, then it seems your model is just not suffering much from training on the noisy labels.

2) Instead of correcting the labels, you could alternatively try deleting the data that cleanlab identified label issues in.
This is what is done by default in our CleanLearning class, because we are not necessarily confident in cleanlab's automatic ability to fix a label (but we are confident in its ability to automatically identify bad labels). 

3) You might consider a different metric than accuracy like auc_roc. The benefits of training on clean data may be more pronounced for more refined metrics, accuracy is quite a crude metric to use for most binary classification tasks. 
Especially if the classes are imbalanced in the original dataset.

**Comment by filippoBUO** [comment]
Thank you for your suggestions @jwmueller, I'll try out!

Good luck!

**Comment by jwmueller** [comment]
Closing issue for now, feel free to reopen if you have more questions
